internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/internal/client/api"
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/li1234yun/gods-generic/maps/linkedhashmap"
+)
+
+const numCachedHeaders = 8
+
+// An extrinsic entry in the database.
+type dbExtrinsic[E runtime.Extrinsic] struct {
+	inner any
+}
+
+type dbExtrinsicValues[E runtime.Extrinsic] interface {
+	dbExtrinsicIndexed | dbExtrinsicFull[E]
+}
+
+func setVote[E runtime.Extrinsic, Value dbExtrinsicValues[E]](mvdt *dbExtrinsic[E], value Value) {
+	mvdt.inner = value
+}
+
+func (mvdt *dbExtrinsic[E]) SetValue(value any) (err error) {
+	switch value := value.(type) {
+	case dbExtrinsicIndexed:
+		setVote[E](mvdt, value)
+		return
+	case dbExtrinsicFull[E]:
+		setVote[E](mvdt, value)
+		return
+	default:
+		return fmt.Errorf("unsupported type")
+	}
+}
+
+func (mvdt dbExtrinsic[E]) IndexValue() (index uint, value any, err error) {
+	switch mvdt.inner.(type) {
+	case dbExtrinsicIndexed:
+		return 0, mvdt.inner, nil
+	case dbExtrinsicFull[E]:
+		return 1, mvdt.inner, nil
+	}
+	return 0, nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+func (mvdt dbExtrinsic[E]) Value() (value any, err error) {
+	_, value, err = mvdt.IndexValue()
+	return
+}
+
+func (mvdt dbExtrinsic[E]) ValueAt(index uint) (value any, err error) {
+	switch index {
+	case 0:
+		return *new(dbExtrinsicIndexed), nil
+	case 1:
+		return *new(dbExtrinsicFull[E]), nil
+	}
+	return nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+// Extrinsic that contains indexed data.
+type dbExtrinsicIndexed struct {
+	// Hash of the indexed part.
+	Hash hash.H256
+	// Extrinsic header.
+	Header []byte
+}
+type dbExtrinsicFull[E runtime.Extrinsic] struct {
+	Extrinsic E
+}
+
+type metaUpdate[H, N any] struct {
+	Hash        H
+	Number      N
+	IsBest      bool
+	IsFinalized bool
+	WithState   bool
+}
+
+// blockchainDB is the block database.
+type blockchainDB[H runtime.Hash, N runtime.Number, E runtime.Extrinsic, Header runtime.Header[N, H]] struct {
+	db                   database.Database[hash.H256]
+	meta                 meta[H, N]
```

Comment:
Should the mtx values be in the appropriate structure themselves rather than alongside them here? For example, putting metaMtx within Meta for better coupling and I would imagine enforcement of locking right?

Reply:
After thinking on this more I actually do not mind how it is, resolving